### PR TITLE
Store unlocked teleporters of a player

### PR DESCRIPTION
**Pull request by maexled**
Part of Gamify-IT/issues#390

### CODE REVIEW

File: .github/workflows/deploy-docker-image-on-release.yml
Comment (delvh) [inline .github/workflows/deploy-docker-image-on-release.yml]: Don't forget to remove this line again.

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
     * Unlock a teleporter for a player.
```

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
     * @return the updated player statistic
```

**Review by Gr33ndev — APPROVED** [review]

**Review by madebyTimo — APPROVED** [review]
nice code :+1:

**Comment by delvh** [comment]
Will try to review it until the Daily tomorrow, don't merge yet.

File: src/test/java/de/unistuttgart/overworldbackend/TeleporterInputControllerTest.java
Comment (delvh) [inline src/test/java/de/unistuttgart/overworldbackend/TeleporterInputControllerTest.java:161]: I recommend also checking that before, there were exactly zero elements.

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatistic.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatistic.java]: Why a list?
Why not a set?
That makes so many things easier here…

File: src/main/java/de/unistuttgart/overworldbackend/controller/TeleporterInputController.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/controller/TeleporterInputController.java]:
```suggestion
        log.debug("request to update unlocked teleporters for course {} with {}", courseId, data);
```

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java]: Shouldn't we also store the course?

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java:20]: I don't think `index` is a good idea.
I prefer letting the frontend send a string, as that will be far more backwards compatible.

File: src/main/java/de/unistuttgart/overworldbackend/data/config/DungeonConfig.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/config/DungeonConfig.java]: ?

File: src/main/java/de/unistuttgart/overworldbackend/data/mapper/TeleporterMapper.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/mapper/TeleporterMapper.java]:
```suggestion
@Mapper(componentModel = "spring", uses = AreaLocationMapper.class)
```

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java
Comment (delvh) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java:20]: Also: Why `@NotNull`? And why suddenly `javax.validation.constraints` instead of `org.springframework.lang`?

File: src/main/java/de/unistuttgart/overworldbackend/data/Position.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/Position.java]: Erm… Wasn't there something that we don't need this?
On top it is even unused.

File: src/main/java/de/unistuttgart/overworldbackend/data/Teleporter.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/Teleporter.java:23]: As mentioned, please convert that to a `NotBlank` string.

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java:133]: Don't forget to replace this with `.equals` then

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java:141]:
```suggestion
                        "Teleporter with name %s in area with id %s is already unlocked.",
```

**Review by delvh — CHANGES_REQUESTED** [review]

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java
Comment (maexled) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java]: The course is submitted by pathvariable

Comment (delvh) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java]: Still applies, but for teleporter instead.

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java
Comment (maexled) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java:20]: After consultation with @PatrickHoltz , we don't think the idea is so good, because we have made it uniform as an index so far and the frontend team gets along so well.

Comment (delvh) [reply src/main/java/de/unistuttgart/overworldbackend/data/PlayerTeleporterData.java:20]: Okay, fine.

File: src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatistic.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/data/PlayerStatistic.java]:
```suggestion
        this.unlockedTeleporters.add(teleporter);
```
It's a set :)

File: src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/PlayerStatisticService.java]:
```suggestion
                        "Teleporter with index %s in area with id %s is already unlocked.",
```

**Review by delvh — APPROVED** [review]
Why is the clonetest failing now?

**Comment by delvh** [comment]
Erm… Did you just merge a branch with a failing CI?

**Pull request merged**
